Outlines.Inspection/AutomationPropertiesWatcher.cs

```diff
@@ -0,0 +1,100 @@
+using System.Threading;
+using System.Threading.Tasks;
+using UIAutomationClient;
+using Outlines.Core;
+
+namespace Outlines.Inspection
+{
+    public class AutomationPropertiesWatcher
+    {
+        private static readonly TimeSpan RefreshRate = TimeSpan.FromMilliseconds(300);
+
+        private IOutlinesService OutlinesService { get; set; }
+        private IElementPropertiesProvider ElementPropertiesProvider { get; set; }
+
+        private IUIAutomationElement WatchedSelectedElement { get; set; }
+        private IUIAutomationElement WatchedTargetElement { get; set; }
+        private CancellationTokenSource WatchSelectedElementTaskCancellationTokenSource { get; set; }
+        private CancellationTokenSource WatchTargetElementTaskCancellationTokenSource { get; set; }
+
+        public AutomationPropertiesWatcher(IOutlinesService outlinesService, IElementPropertiesProvider elementPropertiesProvider)
+        {
+            ElementPropertiesProvider = elementPropertiesProvider;
+            OutlinesService = outlinesService;
+            OutlinesService.SelectedElementChanged += OnSelectedElementChanged;
+            OutlinesService.TargetElementChanged += OnTargetElementChanged;
+        }
+
+        private void OnSelectedElementChanged()
+        {
+            var selectedElementProperties = OutlinesService.SelectedElementProperties as AutomationElementProperties;
+            if (selectedElementProperties != null)
+            {
+                if (!AreSameElement(WatchedSelectedElement, selectedElementProperties.Element))
+                {
+                    // Stop watching the previous selected element.
+                    WatchSelectedElementTaskCancellationTokenSource?.Cancel();
+
+                    WatchedSelectedElement = selectedElementProperties.Element;
+                    WatchSelectedElementTaskCancellationTokenSource = new CancellationTokenSource();
+                    StartWatchingElementProperties(WatchedSelectedElement, OutlinesService.SelectElementWithProperties, WatchSelectedElementTaskCancellationTokenSource.Token);
+                }
+            }
+        }
+
+        private void OnTargetElementChanged()
+        {
+            var targetElementProperties = OutlinesService.TargetElementProperties as AutomationElementProperties;
+            if (targetElementProperties != null)
+            {
+                if (!AreSameElement(WatchedTargetElement, targetElementProperties.Element))
+                {
+                    // Stop watching the previous target element.
+                    WatchTargetElementTaskCancellationTokenSource?.Cancel();
+
+                    WatchedTargetElement = targetElementProperties.Element;
+                    WatchTargetElementTaskCancellationTokenSource = new CancellationTokenSource();
+                    StartWatchingElementProperties(WatchedTargetElement, OutlinesService.TargetElementWithProperties, WatchTargetElementTaskCancellationTokenSource.Token);
```

Comment:
Creating a new thread each time the target element changes is resource intensive, we should revisit this when we explore improving performance around targeting elements.